jubjub/src/lib.rs

```diff
@@ -900,6 +969,321 @@ pub fn batch_normalize<'a>(v: &'a mut [ExtendedPoint]) -> impl Iterator<Item = A
+                v,
+            });
+
+            if p.is_some().into() {
+                let p = p.unwrap().to_curve().mul_by_cofactor();
+
+                if bool::from(!p.is_identity()) {
+                    return SubgroupPoint(p);
+                }
+            }
+        }
+    }
+
+    fn identity() -> Self {
+        SubgroupPoint(ExtendedPoint::identity())
+    }
+
+    fn generator() -> Self {
+        SubgroupPoint(ExtendedPoint::generator())
+    }
+
+    fn is_identity(&self) -> Choice {
+        self.0.is_identity()
+    }
+
+    #[must_use]
+    fn double(&self) -> Self {
+        SubgroupPoint(self.0.double())
+    }
+}
+
+impl WnafGroup for ExtendedPoint {
+    fn recommended_wnaf_for_scalar(_scalar: &Self::Scalar) -> usize {
+        let num_bits = Self::Scalar::NUM_BITS as usize;
+
+        // Copied from bls12_381::g1, should be updated.
+        if num_bits >= 130 {
+            4
+        } else if num_bits >= 34 {
+            3
+        } else {
+            2
+        }
+    }
+
+    fn recommended_wnaf_for_num_scalars(num_scalars: usize) -> usize {
+        // Copied from bls12_381::g1, should be updated.
+        const RECOMMENDATIONS: [usize; 12] =
+            [1, 3, 7, 20, 43, 120, 273, 563, 1630, 3128, 7933, 62569];
+
+        let mut ret = 4;
+        for r in &RECOMMENDATIONS {
+            if num_scalars > *r {
+                ret += 1;
+            } else {
+                break;
+            }
+        }
+
+        ret
+    }
+}
+
+impl PrimeGroup for SubgroupPoint {}
+
+impl CofactorGroup for ExtendedPoint {
+    type Subgroup = SubgroupPoint;
+
+    fn mul_by_cofactor(&self) -> Self::Subgroup {
+        SubgroupPoint(self.mul_by_cofactor())
+    }
+
+    fn into_subgroup(self) -> CtOption<Self::Subgroup> {
+        CtOption::new(SubgroupPoint(self), self.is_torsion_free())
+    }
+}
+
+impl Curve for ExtendedPoint {
+    type AffineRepr = AffinePoint;
+
+    fn batch_normalize(p: &[Self], q: &mut [Self::AffineRepr]) {
+        Self::batch_normalize(p, q);
+    }
+
+    fn to_affine(&self) -> Self::AffineRepr {
+        self.into()
+    }
+}
+
+impl CofactorCurve for ExtendedPoint {
+    type Affine = AffinePoint;
+}
+
+impl CofactorCurveAffine for AffinePoint {
+    type Scalar = Fr;
+    type Curve = ExtendedPoint;
+
+    fn identity() -> Self {
+        Self::identity()
+    }
+
+    fn generator() -> Self {
+        // The point with the lowest positive v-coordinate and positive u-coordinate, that
+        // is prime-order.
+        AffinePoint {
+            u: Fq::from_raw([
+                0x4df7_b7ff_ec7b_eaca,
+                0x2e3e_bb21_fd6c_54ed,
+                0xf1fb_f02d_0fd6_cce6,
+                0x3fd2_814c_43ac_65a6,
+            ]),
+            v: Fq::from_raw([
+                0x0000_0000_0000_0012,
+                0x0000_0000_0000_0000,
+                0x0000_0000_0000_0000,
+                0x0000_0000_0000_0000,
+            ]),
+        }
```

Comment:
@daira is verifying this.

Reply:
It's a little weird (in terms of the API) that the `generator()` method of `CofactorCurveAffine` doesn't actually generate the entire group, but I guess it'd be weird for someone to want such an element?

Reply:
Yeah, I was on the fence about this. Originally `G1Projective::random` etc returned only subgroup points, but then I encountered a test that relief in it sampling the full group, so I switched to each group sampling from their full set of group elements. There's a similar argument here, and the weirdness is in terms of what people then expect the non-generic APIs to return.

Reply:
In a pairing with @ebfull, we decided that `Group::generator` should generate the entire group, and that `CofactorGroup::generator().clear_cofactor() == CofactorGroup::Subgroup::generator()`. This means:
- We remove the direct `[G1,G2]Projective::generator` APIs, as they are currently incorrect.
- We `impl Group::generator for [G1,G2]Subgroup` to return their current generators.
- We `impl Group::generator for [G1,G2]Projective` to return `unimplemented!()` for now (to unblock this PR while we reverse-engineer what they should be).
- We switch all `G1Projective::generator` calls to `G1Subgroup::generator` for both type safety and so they continue to work.
- We define Jubjub's generator as above, but without the prime-order restriction, and then use its `clear_by_cofactor` which will be coincident with `mul_by_cofactor` as the cofactor is 8.